src/main/java/org/cubedb/core/MultiCubeImpl.java

```diff
@@ -46,7 +46,9 @@ private Cube createNewCube(String cubeName) {
 
 	@Override
 	public void insert(List<DataRow> data) {
-		Lock l = lock.writeLock();
+		Lock l = lock.readLock(); // Although we are technically writing here,
```

Comment:
Citing the ReadWriteLock docs: "All ReadWriteLock implementations must guarantee that the memory synchronization effects of writeLock operations (as specified in the Lock interface) also hold with respect to the associated readLock. That is, a thread successfully acquiring the read lock will see all updates made upon previous release of the write lock."

I.e. save/get should hold the read lock, load/insert should hold the write lock.  

Why do you break this contract? It's either I misunderstand something, or there is a serious error here. 

BTW, I suspect it's really hard to catch the error in tests here.